File: web/public/lib/catalog/series-summary.js
/**
 * What a show adds up to, from the episodes the catalog already carries.
 *
 * Nothing here is fetched or stored: runtime, size, picture and languages are
 * all facts about the files themselves, so a show describes itself as soon as
 * one episode of it exists. That also means every field can be missing, and a
 * field missing everywhere is left out rather than shown empty — a blank row
 * reads as "we looked and there is none", which would be a lie.
 *
 * Where episodes disagree the range is shown. A show half in 720p and half in
 * 1080p is worth knowing about, and picking either one to display would hide
 * exactly the thing worth seeing.
 */

import { countOf, humanDuration, humanSize } from "../format.js";
import { languageLabel } from "../language-label.js";
import { flattenCollection } from "../library.js";

/**
 * @typedef {import("../../../src/catalog/shows.ts").ShowMeta} ShowMeta
 * @typedef {Partial<Pick<ShowMeta, "totalEpisodes"|"totalSeasons"|"firstAir"|"lastAir">>} SeriesMetadata
 */

/** The parts that have something to say, as one line, or `null` if none do. */
function joined(...parts) {
  return parts.filter(Boolean).join(" \u00b7 ") || null;
}

/** The distinct values of `pick` across `sets`, in first-seen order. */
function distinct(sets, pick) {
  // A Set keeps first-insertion order, which is the order promised above.
  const seen = new Set();
  for (const set of sets) {
    const value = pick(set);
    for (const item of Array.isArray(value) ? value : [value]) {
      if (item !== null && item !== undefined && item !== "") seen.add(item);
    }
  }
  return [...seen];
}

/**
 * Languages from one of the index's JSON list columns; a malformed one is no
 * languages rather than a thrown error.
 *
 * Both audio and subtitles are read from the file's own tracks. The `assets`
 * table knows about extracted subtitle sidecars, which is a different
 * question — what the player can serve as a separate track — and answering
 * the header with it would say a show has no subtitles when its files do.
 */
function languagesIn(value) {
  if (Array.isArray(value)) return value;
  try {
    const parsed = JSON.parse(value ?? "[]");
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

/** Quality labels smallest first, so a range reads the way a viewer expects. */
const QUALITY_ORDER = ["SD", "480p", "720p", "1080p", "1440p", "2160p"];

function sortedQualities(values) {
  return [...values].sort((a, b) => {
    const ai = QUALITY_ORDER.indexOf(a);
    const bi = QUALITY_ORDER.indexOf(b);
    // An unknown label sorts after the known ones rather than to the front.
    return (ai === -1 ? QUALITY_ORDER.length : ai) - (bi === -1 ? QUALITY_ORDER.length : bi);
  });
}

/**
 * The facts a show's header states.
 *
 * Returned as data rather than markup so it can be tested without a DOM, the
 * way the rest of this directory's grouping already is.
 * @param {Pick<import("../library.js").Collection, "divisions">} collection
 */
export function summarize(collection) {
  const sets = flattenCollection(collection);
  const seasons = collection.divisions?.length ?? 0;

  const years = distinct(sets, (s) => s.year).sort((a, b) => a - b);
  const qualities = sortedQualities(distinct(sets, (s) => s.quality));
  // SDR is the absence of anything worth saying, so it is only worth saying
  // when some episodes have more than it.
  const hdr = distinct(sets, (s) => s.hdr).filter((v) => v !== "SDR");

  const runtime = sets.reduce((n, s) => n + (Number(s.duration) || 0), 0);
  const size = sets.reduce((n, s) => n + (Number(s.total) || 0), 0);

  return {
    episodes: sets.length,
    seasons,
    runtime: runtime > 0 ? runtime : null,
    size: size > 0 ? size : null,
    years,
    qualities,
    hdr,
    video: distinct(sets, (s) => s.vcodec),
    audio: distinct(sets, (s) => s.acodec),
    audioLanguages: distinct(sets, (s) => languagesIn(s.alang)).map((code) => languageLabel(code, code)),
    subtitleLanguages: distinct(sets, (s) => languagesIn(s.slang)).map((code) => languageLabel(code, code)),
  };
}

/** `1080p`, or `720p–1080p` when a show is not all one thing. */
export function rangeOf(values) {
  if (values.length === 0) return null;
  if (values.length === 1) return values[0];
  return `${values[0]}–${values[values.length - 1]}`;
}

/** `1080p · HDR10 · h264 · aac`, or `null` when nothing was recorded. */
export function pictureLine(facts) {
  return joined(rangeOf(facts.qualities), ...facts.hdr, ...facts.video, ...facts.audio);
}

/**
 * How much of a show is held: `8 episodes · one season · 5h 58m · 38.7 GB`.
 *
 * When the provider has said how much exists and this library has less, the
 * counts say so — `8 of 16 episodes`. Holding all of it says nothing about
 * totals, because "16 of 16" is a fact about arithmetic, not about the show.
 * @param {SeriesMetadata|null} [meta]
 */
export function scaleLine(facts, meta = null) {
  return joined(
    held(facts.episodes, meta?.totalEpisodes ?? null, "episode"),
    facts.seasons > 0 ? held(facts.seasons, meta?.totalSeasons ?? null, "season") : null,
    facts.runtime === null ? null : humanDuration(facts.runtime),
    facts.size === null ? null : humanSize(facts.size),
  );
}

/**
 * `one season`, or `1 of 2 seasons` when some of it is missing.
 *
 * Numerals once a comparison is involved: `one of two seasons` reads as prose
 * where the point is the arithmetic.
 */
function held(have, total, noun) {
  if (typeof total !== "number" || total <= have) return countOf(have, noun);
  return `${have} of ${total} ${noun}s`;
}

/**
 * The labelled rows worth printing, in order.
 *
 * A language nobody recorded yields no row at all: an empty `Audio` line
 * would say the library looked and found none, which is not what happened.
 */
export function detailRows(facts) {
  return [
    { label: "Audio", value: facts.audioLanguages.join(", ") },
    { label: "Subtitles", value: facts.subtitleLanguages.join(", ") },
  ].filter((row) => row.value !== "");
}

/**
 * The years a show ran, or failing that the years this library's copies carry.
 *
 * The provider's dates describe the show; the episodes' years describe what is
 * held. A library with one season of a show that ran seven should say when the
 * show ran, so the dates win where they exist.
 * @param {SeriesMetadata|null} [meta]
 */
export function yearLine(facts, meta = null) {
  const first = year(meta?.firstAir);
  if (first !== null) {
    const last = year(meta?.lastAir);
    // A show still running has no last date, and an en dash to nowhere would
    // claim it ended.
    return last !== null && last !== first ? `${first}\u2013${last}` : String(first);
  }
  return facts.years.length > 0 ? rangeOf(facts.years.map(String)) : null;
}

/** The year out of a provider date, which is `YYYY-MM-DD` or absent. */
function year(date) {
  const parsed = Number(String(date ?? "").slice(0, 4));
  return Number.isInteger(parsed) && parsed > 1800 ? parsed : null;
}

/**
 * Genre, rating, network and status on one line, from the provider.
 *
 * A rating of zero is what an unrated title scores and is not a rating, so
 * the uploader records nothing for it and nothing is printed.
 * @param {Partial<Pick<ShowMeta, "genres"|"rating"|"network"|"status">>|null} meta
 */
export function provenance(meta) {
  if (!meta) return null;
  return joined(
    meta.genres,
    typeof meta.rating === "number" ? `\u2605 ${meta.rating.toFixed(1)}` : null,
    meta.network,
    meta.status,
  );
}

File: android/feature/catalog/src/main/kotlin/SeriesSummary.kt
package catalog

import model.MediaSet
import model.humanSize
import uniffi.mediagram_core.TitleInfo

/**
 * What a show adds up to, from the episodes the catalog already holds —
 * ported from the web's `series-summary.js#summarize`, narrowed to what
 * [TitleInfo] actually carries on Android: no first/last air date and no
 * provider episode/season totals reach this build yet (`TitleInfo` has
 * `overview`/`tagline`/`genres`/`rating`/`network`/`status` and nothing
 * else), so "Aired" and "Held" can only speak from the episodes this
 * library holds, never as a count against a total the provider named.
 */
data class SeriesFacts(
    val episodes: Int,
    val seasons: Int,
    val runtimeSecs: Int,
    val totalBytes: Long,
    val years: List<Int>,
    val qualities: List<String>,
    val hdr: List<String>,
    val video: List<String>,
    val audio: List<String>,
    val subtitleLanguages: List<String>,
)

private val QUALITY_ORDER = listOf("SD", "480p", "720p", "1080p", "1440p", "2160p")

/** The distinct values [pick] names across [sets], first-seen order kept — mirrors `series-summary.js`'s own `distinct`. */
private fun <T> distinct(
    sets: List<MediaSet>,
    pick: (MediaSet) -> T?,
): List<T> {
    val seen = LinkedHashSet<T>()
    for (set in sets) pick(set)?.let(seen::add)
    return seen.toList()
}

private fun <T> distinctAll(
    sets: List<MediaSet>,
    pick: (MediaSet) -> List<T>,
): List<T> {
    val seen = LinkedHashSet<T>()
    for (set in sets) seen.addAll(pick(set))
    return seen.toList()
}

fun summarize(divisions: List<Division>): SeriesFacts {
    val sets = divisions.flatMap { it.walk() }.flatMap { it.items }
    val qualities =
        distinct(sets) { it.quality }.sortedBy { q ->
            QUALITY_ORDER.indexOf(q).let { if (it < 0) QUALITY_ORDER.size else it }
        }
    return SeriesFacts(
        episodes = sets.size,
        seasons = divisions.size,
        runtimeSecs = sets.sumOf { it.durationSecs ?: 0 },
        totalBytes = sets.sumOf { it.totalBytes },
        years = distinct(sets) { it.year?.takeIf { y -> y > 0 } }.sorted(),
        qualities = qualities,
        // SDR is the absence of anything worth saying, so it only counts once some episode has more than it.
        hdr = distinct(sets) { it.hdr?.takeIf { h -> h != "SDR" } },
        video = distinct(sets) { it.vcodec },
        audio = distinct(sets) { it.acodec },
        subtitleLanguages = distinctAll(sets) { it.subtitleLanguages },
    )
}

/** `1080p`, or `720p–1080p` when a show is not all one thing. */
fun rangeOf(values: List<String>): String? =
    when {
        values.isEmpty() -> null
        values.size == 1 -> values[0]
        else -> "${values.first()}–${values.last()}"
    }

private fun joinedOf(vararg parts: String?): String? =
    parts.filterNotNull().filter { it.isNotEmpty() }.joinToString(" · ").takeIf { it.isNotEmpty() }

/** `1080p · HDR10 · h264 · aac`, or `null` when nothing was recorded. */
fun pictureLine(facts: SeriesFacts): String? =
    joinedOf(rangeOf(facts.qualities), *facts.hdr.toTypedArray(), *facts.video.toTypedArray(), *facts.audio.toTypedArray())

/**
 * How much of a show is held: `8 episodes · 2 seasons · 5h 58m · 38.7 GB`.
 *
 * Always what this library has, never "8 of 16" — [TitleInfo] carries no
 * provider total to compare against (see this file's own doc comment).
 */
fun scaleLine(facts: SeriesFacts): String? =
    joinedOf(
        held(facts.episodes, "episode"),
        facts.seasons.takeIf { it > 0 }?.let { held(it, "season") },
        humanDuration(facts.runtimeSecs),
        facts.totalBytes.takeIf { it > 0 }?.let(::humanSize),
    )

private fun held(
    count: Int,
    noun: String,
): String = "$count $noun${if (count == 1) "" else "s"}"

/** The years a show ran, from the episodes this library holds — no provider air dates reach this build (see this file's own doc comment). */
fun yearLine(facts: SeriesFacts): String? = facts.years.takeIf { it.isNotEmpty() }?.let { rangeOf(it.map(Int::toString)) }

/**
 * Rating, network and status from the provider, on one line. Genres are
 * their own row in the About tab's fact sheet, so they are not repeated
 * here — mirrors `series-page.js` calling `provenance` with `genres: null`.
 */
fun provenance(info: TitleInfo?): String? = info?.let { joinedOf(ratingLabel(it.rating), it.network, it.status) }

/**
 * The one row [pictureLine] leaves out: subtitle languages. There is no
 * "Audio languages" row to match `series-summary.js`'s own `detailRows` —
 * [model.MediaSet] carries no audio-language field yet, only [model.MediaSet.subtitleLanguages].
 */
fun detailRows(facts: SeriesFacts): List<Pair<String, String>> =
    listOfNotNull(
        facts.subtitleLanguages.takeIf { it.isNotEmpty() }
            ?.let { codes -> "Subtitles" to codes.joinToString(", ", transform = ::languageName) },
    )

/** A language code as a viewer reads it: `en` → `English`. Mirrors the web's `languageLabel`, minus its own dictionary — `Locale`'s carries the same names. */
fun languageName(code: String): String =
    java.util.Locale.forLanguageTag(code).getDisplayLanguage(java.util.Locale.ENGLISH).ifEmpty { code }
